pkg/controller/servicing/lifecycler.go

```diff
@@ -0,0 +1,359 @@
+	}
+	return nil
+}
+
+// Replace a node by temrinating it
+func (lc *NodeLifeCycler) Replace(node *core_v1.Node) error {
+	if err := lc.Openstack.DeleteNode(node.Spec.ExternalID); err != nil {
+		return errors.Wrap(err, "deleting node failed")
+	}
+	return nil
+}
+
+// Drain logs the action
+func (lc *LoggingLifeCycler) Drain(node *core_v1.Node) (err error) {
+	defer func(begin time.Time) {
+		lc.Logger.Log(
+			"msg", "draining node",
+			"node", node.GetName(),
+			"took", time.Since(begin),
+			"v", 1,
+			"err", err,
+		)
+	}(time.Now())
+	return lc.LifeCycler.Drain(node)
+}
+
+// Reboot logs the action
+func (lc *LoggingLifeCycler) Reboot(node *core_v1.Node) (err error) {
+	defer func(begin time.Time) {
+		lc.Logger.Log(
+			"msg", "rebooting node",
+			"node", node.GetName(),
+			"took", time.Since(begin),
+			"v", 1,
+			"err", err,
+		)
+	}(time.Now())
+	return lc.LifeCycler.Reboot(node)
+}
+
+// Replace logs the action
+func (lc *LoggingLifeCycler) Replace(node *core_v1.Node) (err error) {
+	defer func(begin time.Time) {
+		lc.Logger.Log(
+			"msg", "replacing node",
+			"node", node.GetName(),
+			"took", time.Since(begin),
+			"v", 1,
+			"err", err,
+		)
+	}(time.Now())
+	return lc.LifeCycler.Replace(node)
+}
+
+// Drain writes an Event
+func (lc *EventingLifeCycler) Drain(node *core_v1.Node) error {
+	err := lc.LifeCycler.Drain(node)
+	if err == nil {
+		lc.Recorder.Eventf(
+			lc.Kluster,
+			core_v1.EventTypeNormal,
+			events.SuccessfulDrainNode,
+			"Preparing upgrade for node: %v. Successfully drained node.",
+			node.GetName())
+	} else {
+		lc.Recorder.Eventf(
+			lc.Kluster,
+			core_v1.EventTypeNormal,
+			events.FailedDrainNode,
+			"Preparing upgrade for node: %v. Failed to drain node: %v",
+			node.GetName(),
+			err)
+	}
+	return err
+}
+
+// Reboot writes an Event
+func (lc *EventingLifeCycler) Reboot(node *core_v1.Node) error {
+	err := lc.LifeCycler.Reboot(node)
+	if err == nil {
+		lc.Recorder.Eventf(
+			lc.Kluster,
+			core_v1.EventTypeNormal,
+			events.SuccessfulRebootNode,
+			"Upgrading OS for node: %v. Reboot successful.",
+			node.GetName())
+	} else {
+		lc.Recorder.Eventf(
+			lc.Kluster,
+			core_v1.EventTypeNormal,
+			events.FailedRebootNode,
+			"Upgrading OS for node: %v. Reboot failed: %v",
+			node.GetName(),
+			err)
+	}
+	return err
+}
+
+// Replace writes an Event
+func (lc *EventingLifeCycler) Replace(node *core_v1.Node) error {
+	err := lc.LifeCycler.Reboot(node)
+	if err == nil {
+		lc.Recorder.Eventf(
+			lc.Kluster,
+			core_v1.EventTypeNormal,
+			events.SuccessfulReplaceNode,
+			"Replacing node for upgrade: %v. Termination successful",
+			node.GetName())
+	} else {
+		lc.Recorder.Eventf(
+			lc.Kluster,
+			core_v1.EventTypeNormal,
+			events.FailedReplaceNode,
+			"Replacing node for upgrade: %v. Termination failed: %v",
+			node.GetName(),
+			err)
+	}
+	return err
+}
+
+// Drain collects metrics
+func (lc *InstrumentingLifeCycler) Drain(node *core_v1.Node) (err error) {
```

Comment:
the `prometheus.Labels` struct could be defined once in the beginning and passed to each `With` method. maybe even in the `InstrumentingLifeCycler` struct and re-used in all the instrumenting functions.